File: src/agent.js
/**
 * Main agent logic.
 * @module agent
 */

import readline from 'readline';
import { config } from './config.js';
import { callLLM } from './llm.js';
import { tools, toolImplementations } from './tools.js';

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

const systemPrompt = `You are a powerful agentic AI assistant.
You have access to a bash tool which allows you to do almost anything on the system.
You should use this tool to accomplish the user's requests.
You are optimized for high reasoning and complex tasks.
Always verify your actions and output.
If you need to run a command, just do it.
The user has set a strict output limit of 1k tokens per tool call. If you see truncated output, refine your command (e.g., use grep, head, tail) to get the specific information you need.`;

let messages = [
  { role: "system", content: systemPrompt },
];

let totalCost = 0;
let hasSeenCachedTokens = false;
let lastRequestTime = 0;

/**
 * Asks for user approval.
 * @param {string} question - The question to ask.
 * @returns {Promise<boolean>} True if approved.
 */
function askApproval(question) {
  return new Promise((resolve) => {
    rl.question(`${question} (y/n): `, (answer) => {
      resolve(answer.toLowerCase() === 'y');
    });
  });
}

/**
 * Main agent loop.
 */
export async function startAgent() {
  console.log(`Agent started in ${config.mode} mode using model ${config.model}`);
  console.log("Type 'exit' to quit.");

  const askUser = () => {
    rl.question('User: ', async (input) => {
      if (input.toLowerCase() === 'exit') {
        rl.close();
        return;
      }

      messages.push({ role: "user", content: input });

      await processTurn();
      
      askUser();
    });
  };

  askUser();
}

/**
 * Processes a single turn of the agent.
 */
async function processTurn() {
  let turnFinished = false;
  
  while (!turnFinished) {
    console.log("Thinking...");
    try {
      // Manage cache checkpoints before calling LLM
      manageCache(messages);

      
      const currentTime = Date.now();
      let elapsedMinutes = 0;
      if (lastRequestTime > 0) {
        elapsedMinutes = (currentTime - lastRequestTime) / 60000;
      }
      const { message: responseMessage, usage } = await callLLM(messages, tools);
      lastRequestTime = Date.now();
      
      if (usage) {
        if (usage.cost) {
          totalCost += usage.cost;
          console.log(`Total Session Cost: $${totalCost.toFixed(6)}`);
        }

        const cachedTokens = usage.prompt_tokens_details?.cached_tokens || usage.cachedContentTokenCount || 0;
        if (cachedTokens > 0) {
          hasSeenCachedTokens = true;
        }

        if (hasSeenCachedTokens && cachedTokens === 0) {
          // Gemini cache TTL is 60 minutes (default), Anthropic is 5 minutes
          const isGemini = config.provider === 'gemini' || config.model.includes('gemini');
          const cacheTTL = isGemini ? 60.0 : 5.0;
          const reason = elapsedMinutes < (cacheTTL - 1) ? "Prefix mismatch or Checkpoint limit" : "Cache TTL expired";
          console.log(`\x1b[31mWARNING: Cached tokens dropped to 0! (Elapsed: ${elapsedMinutes.toFixed(1)} minutes). Cause: ${reason}.\x1b[0m`);
        }
      }
      
      // Add assistant message to history
      messages.push(responseMessage);

      if (responseMessage.tool_calls) {
        for (const toolCall of responseMessage.tool_calls) {
          const functionName = toolCall.function.name;
          const args = JSON.parse(toolCall.function.arguments);
          const toolFunc = toolImplementations[functionName];

          if (toolFunc) {
            console.log(`Tool Call: ${functionName}(${JSON.stringify(args)})`);
            
            let approved = true;
            if (config.mode === 'manual') {
              approved = await askApproval("Execute this command?");
            }

            let result;
            if (approved) {
              console.log("Executing...");
              result = await toolFunc(args);
            } else {
              result = "User denied execution of this command.";
            }

            console.log(`Result: ${result.substring(0, 100)}...`); // Log brief result

            messages.push({
              role: "tool",
              tool_call_id: toolCall.id,
              name: functionName,
              content: result,
            });
          }
        }
        // Loop continues to let the model respond to the tool output
      } else {
        console.log(`Agent: ${responseMessage.content}`);
        turnFinished = true;
      }
    } catch (error) {
      console.error("\n\x1b[31mError during agent execution:\x1b[0m", error.message);
      console.log("The current turn has been aborted due to an error. You can try again or enter a new command.");
      turnFinished = true;
    }
  }
}

/**
 * Manages cache control headers in messages.
 * Adds explicit cache checkpoints to the system prompt and periodically in the history.
 * Implements a rolling window of checkpoints to stay within the 4-checkpoint limit.
 * @param {Array} messages 
 */
function manageCache(messages) {
  const isAnthropic = config.model.includes('anthropic') || config.model.includes('claude');
  const isGemini = config.model.includes('gemini');
  
  // Only apply caching for supported models
  if (!isAnthropic && !isGemini) {
    return;
  }

  // 1. Ensure System Prompt has cache_control
  const systemMsg = messages.find(m => m.role === 'system');
  if (systemMsg) {
    if (typeof systemMsg.content === 'string') {
      systemMsg.content = [{ type: "text", text: systemMsg.content, cache_control: { type: "ephemeral" } }];
    } else if (Array.isArray(systemMsg.content)) {
      // Check if already has cache_control
      const hasCache = systemMsg.content.some(block => block.cache_control);
      if (!hasCache && systemMsg.content.length > 0) {
        // Add to the last block
        systemMsg.content[systemMsg.content.length - 1].cache_control = { type: "ephemeral" };
      }
    }
  }

  // 2. Manage History Checkpoints
  // Gemini: Only uses the LAST cache_control breakpoint (OpenRouter limitation)
  // Anthropic: Can use up to 4 checkpoints total
  
  if (isGemini) {
    // For Gemini: Only add cache_control to the most recent cacheable message
    // This ensures we cache the longest possible prefix
    // Remove all existing cache_control from non-system messages first
    for (let i = 0; i < messages.length; i++) {
      const msg = messages[i];
      if (msg.role === 'system') continue;
      
      if (Array.isArray(msg.content)) {
        msg.content.forEach(block => {
          if (block.cache_control) delete block.cache_control;
        });
      }
    }
    
    // Find the last cacheable message (must have content)
    let lastCacheableIndex = -1;
    for (let i = messages.length - 1; i >= 0; i--) {
      const msg = messages[i];
      if (msg.role === 'system') continue;
      
      const hasContent = msg.content && (typeof msg.content === 'string' || (Array.isArray(msg.content) && msg.content.length > 0));
      if (hasContent) {
        lastCacheableIndex = i;
        break;
      }
    }
    
    // Add cache_control to the last cacheable message
    if (lastCacheableIndex !== -1) {
      const msg = messages[lastCacheableIndex];
      if (typeof msg.content === 'string') {
        msg.content = [{ type: "text", text: msg.content, cache_control: { type: "ephemeral" } }];
      } else if (Array.isArray(msg.content) && msg.content.length > 0) {
        msg.content[msg.content.length - 1].cache_control = { type: "ephemeral" };
      }
      console.log(`\x1b[32m[Cache] Gemini checkpoint set at message ${lastCacheableIndex} (Role: ${msg.role})\x1b[0m`);
    }
    
  } else if (isAnthropic) {
    // Anthropic: Use rolling window with multiple checkpoints
    const SYSTEM_AND_TOOLS_CHECKPOINTS = 2;
    const MAX_CHECKPOINTS = 4;
    const HISTORY_CHECKPOINTS_QUOTA = MAX_CHECKPOINTS - SYSTEM_AND_TOOLS_CHECKPOINTS; // 2
    const CHECKPOINT_INTERVAL = 8; 

    // Calculate desired checkpoint indices based on interval
    const candidateIndices = [];

    // Identify potential candidates (Every Nth message)
    for (let i = 0; i < messages.length; i++) {
      const msg = messages[i];
      if (msg.role === 'system') continue;

      if (i % CHECKPOINT_INTERVAL === 0 && i > 0) {
        candidateIndices.push(i);
      }
    }

    // Resolve candidates to actual valid cacheable messages
    let finalIndices = [];
    
    for (const index of candidateIndices) {
       let bestIndex = -1;
       const searchOrder = [index, index - 1, index + 1, index - 2, index + 2];
       
       for (const searchIdx of searchOrder) {
         if (searchIdx > 0 && searchIdx < messages.length) {
           const m = messages[searchIdx];
           const hasContent = m.content && (typeof m.content === 'string' || (Array.isArray(m.content) && m.content.length > 0));
           if (hasContent) {
             bestIndex = searchIdx;
             break;
           }
         }
       }

       if (bestIndex !== -1) {
         if (!finalIndices.includes(bestIndex)) {
           finalIndices.push(bestIndex);
         }
       }
    }

    // Keep only the last N desired indices
    if (finalIndices.length > HISTORY_CHECKPOINTS_QUOTA) {
      finalIndices = finalIndices.slice(-HISTORY_CHECKPOINTS_QUOTA);
    }

    // Apply changes to messages
    for (let i = 0; i < messages.length; i++) {
      const msg = messages[i];
      if (msg.role === 'system') continue;

      const isDesired = finalIndices.includes(i);
      let hasCheckpoint = false;
      
      if (Array.isArray(msg.content)) {
        hasCheckpoint = msg.content.some(block => block.cache_control);
      } else if (typeof msg.content === 'string') {
        hasCheckpoint = false;
      }

      if (hasCheckpoint && !isDesired) {
        // Remove checkpoint
        if (Array.isArray(msg.content)) {
          msg.content.forEach(block => {
            if (block.cache_control) delete block.cache_control;
          });
        }
        console.log(`\x1b[33m[Cache] Checkpoint removed at message ${i}\x1b[0m`);
      } else if (!hasCheckpoint && isDesired) {
        // Add checkpoint
        if (typeof msg.content === 'string') {
          msg.content = [{ type: "text", text: msg.content }];
        }
        
        if (Array.isArray(msg.content) && msg.content.length > 0) {
          msg.content[msg.content.length - 1].cache_control = { type: "ephemeral" };
          console.log(`\x1b[32m[Cache] Checkpoint added at message ${i} (Role: ${msg.role})\x1b[0m`);
        }
      }
    }
  }
}

File: src/agent.py
"""Main agent loop and UI."""

import asyncio

try:
    import uvloop

    asyncio.set_event_loop_policy(uvloop.EventLoopPolicy())
except ImportError:
    pass

import os
import json
import time
import logging
import html
from prompt_toolkit import PromptSession, print_formatted_text
from prompt_toolkit.key_binding import KeyBindings
from prompt_toolkit.formatted_text import HTML, FormattedText
from prompt_toolkit.styles import Style
from prompt_toolkit.lexers import PygmentsLexer
from prompt_toolkit.application import Application
from prompt_toolkit.layout.containers import Window
from prompt_toolkit.layout.controls import FormattedTextControl
from prompt_toolkit.layout.layout import Layout
from prompt_toolkit.patch_stdout import patch_stdout
from prompt_toolkit.key_binding.vi_state import InputMode
from pygments.lexers.markup import MarkdownLexer

from .config import config
from .llm import call_llm
from .tools import TOOLS, TOOL_SCHEMAS, save_clipboard_image
from .undo import UndoManager

# ---------------------------------------------------------------------------
# Setup
# ---------------------------------------------------------------------------

# Restore working directory if provided
if "AGENT_WORK_DIR" in os.environ:
    os.chdir(os.environ["AGENT_WORK_DIR"])


class PTKHandler(logging.Handler):
    """Logging handler that uses prompt_toolkit for output."""

    def emit(self, record):
        try:
            msg = html.escape(self.format(record))
            color = {logging.ERROR: "ansired", logging.WARNING: "ansiyellow"}.get(record.levelno, "ansiwhite")
            print_formatted_text(HTML(f"<{color}>{msg}</{color}>"))
        except Exception:
            self.handleError(record)


# Configure logging
logging.getLogger().handlers = []
handler = PTKHandler()
handler.setFormatter(logging.Formatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s"))
logging.getLogger().addHandler(handler)
logging.getLogger().setLevel(logging.DEBUG if config.debug else logging.INFO)
logger = logging.getLogger("Agent")

# ---------------------------------------------------------------------------
# System Prompt
# ---------------------------------------------------------------------------

SYSTEM_PROMPT = """You are a powerful agentic AI assistant.
You have access to a bash tool which allows you to do almost anything on the system.
You should use this tool to accomplish the user's requests.
You are optimized for high reasoning and complex tasks.
Always verify your actions and output.
If you need to run a command, just do it.
The user has set a strict output limit of 1k tokens per tool call. If you see truncated output, refine your command (e.g., use grep, head, tail) to get the specific information you need.

String and scalar parameters should be specified as is, while lists and objects should use JSON format.

Answer the user's request using the relevant tool(s), if they are available. Check that all the required parameters for each tool call are provided or can reasonably be inferred from context. IF there are no relevant tools or there are missing values for required parameters, ask the user to supply these values; otherwise proceed with the tool calls. If the user provides a specific value for a parameter (for example provided in quotes), make sure to use that value EXACTLY. DO NOT make up values for or ask about optional parameters.

If you intend to call multiple tools and there are no dependencies between the calls, make all of the independent calls in the same block, otherwise you MUST wait for previous calls to finish first to determine the dependent values (do NOT use placeholders or guess missing parameters).

Guidelines for tool usage:
- For bash commands, prefer concise commands that get specific information
- Use grep, head, tail, awk to filter output when needed
- For file operations, verify paths exist before modifying
- When searching code, use grep or find to locate relevant files first
- Always check command exit status and handle errors appropriately
- For complex multi-step tasks, break them down and verify each step

When working with files:
- Read files before modifying to understand context
- Use partial updates (old_content parameter) when possible for precision
- Create backups of important files before major changes
- Verify changes after making them

For debugging and investigation:
- Start with broad searches, then narrow down
- Check logs and error messages carefully
- Test hypotheses incrementally
- Document findings as you go"""

# ---------------------------------------------------------------------------
# State
# ---------------------------------------------------------------------------

messages = [{"role": "system", "content": SYSTEM_PROMPT}]
undo_manager = UndoManager()
total_cost = 0.0
last_request_time = 0.0
has_seen_cached_tokens = False

# Wrap update_file to track changes for undo
_original_update_file = TOOLS["update_file"]


async def _tracked_update_file(path, content, old_content=None):
    undo_manager.record_file_change(path)
    return await _original_update_file(path, content, old_content)


TOOLS["update_file"] = _tracked_update_file

# ---------------------------------------------------------------------------
# Core Logic
# ---------------------------------------------------------------------------


def display_usage(usage: dict, elapsed_minutes: float):
    """Display cost and cache information."""
    global total_cost, has_seen_cached_tokens

    cost = usage.get("cost", 0)
    if cost:
        total_cost += cost
        print_formatted_text(HTML(f"<ansicyan>Cost: ${cost:.6f} | Total: ${total_cost:.6f}</ansicyan>"))

    cached = (
        usage.get("cachedContentTokenCount", 0)
        or usage.get("cache_read_input_tokens", 0)
        or usage.get("prompt_tokens_details", {}).get("cached_tokens", 0)
        or 0
    )
    created = usage.get("cache_creation_input_tokens", 0)

    if cached or created:
        print_formatted_text(HTML(f"<ansigreen>Cache: {cached} read, {created} created</ansigreen>"))

    if cached > 0:
        has_seen_cached_tokens = True
    elif has_seen_cached_tokens:
        ttl = 60.0 if config.provider == "gemini" else 5.0
        reason = "Prefix mismatch" if elapsed_minutes < ttl - 1 else "Cache TTL expired"
        print_formatted_text(HTML(f"<ansired>WARNING: Cache dropped to 0! ({reason})</ansired>"))


async def execute_tools(tool_calls: list):
    """Execute tool calls and add results to messages."""
    print_formatted_text(HTML("\n<ansiyellow>Tool Calls:</ansiyellow>"))

    # In manual mode, ask for approval before executing tools
    if config.mode == "manual":
        for tc in tool_calls:
            name = tc["function"]["name"]
            args = json.loads(tc["function"]["arguments"]) if tc["function"]["arguments"] else {}
            print_formatted_text(FormattedText([("bold", f"  {name}"), ("", f"({json.dumps(args)})")]))

        try:
            approval_session = PromptSession()
            response = await approval_session.prompt_async(
                HTML("<ansicyan>Execute? [Y/n]: </ansicyan>")
            )
            response = response.strip().lower()
            if response in ("n", "no"):
                print_formatted_text(HTML("<ansiyellow>Tool execution cancelled.</ansiyellow>"))
                for tc in tool_calls:
                    messages.append({
                        "role": "tool",
                        "tool_call_id": tc.get("id", "unknown"),
                        "name": tc["function"]["name"],
                        "content": "Tool execution cancelled by user."
                    })
                return
        except (EOFError, KeyboardInterrupt):
            print_formatted_text(HTML("\n<ansiyellow>Tool execution cancelled.</ansiyellow>"))
            for tc in tool_calls:
                messages.append({
                    "role": "tool",
                    "tool_call_id": tc.get("id", "unknown"),
                    "name": tc["function"]["name"],
                    "content": "Tool execution cancelled by user."
                })
            return

    for tc in tool_calls:
        name = tc["function"]["name"]
        args = json.loads(tc["function"]["arguments"]) if tc["function"]["arguments"] else {}

        if config.mode == "auto":
            print_formatted_text(FormattedText([("bold", f"  {name}"), ("", f"({json.dumps(args)})")]))
        logger.debug(f"Executing {name} with {args}")

        if name in TOOLS:
            try:
                result = await TOOLS[name](**args)
            except Exception as e:
                result = f"Error: {e}"
                logger.error(f"Tool error: {e}")
        else:
            result = f"Error: Tool '{name}' not found."

        messages.append({"role": "tool", "tool_call_id": tc.get("id", "unknown"), "name": name, "content": str(result)})
        print_formatted_text(FormattedText([("#888888", f"Result: {str(result)[:100]}...")]))


async def process_turn_logic(user_input: str, stop_check=None):
    """Process a single turn of conversation."""
    global last_request_time

    undo_manager.start_turn(messages)
    messages.append({"role": "user", "content": user_input})

    while True:
        if stop_check and stop_check():
            print_formatted_text(HTML("\n<ansiyellow>Stopping after current step.</ansiyellow>"))
            break

        request_time = time.time()
        elapsed = (request_time - last_request_time) / 60.0 if last_request_time else 0
        last_request_time = request_time

        try:
            response = await call_llm(messages, TOOL_SCHEMAS)
        except asyncio.CancelledError:
            raise
        except Exception as e:
            logger.exception("LLM call failed")
            print_formatted_text(HTML(f"\n<ansired>Error: {e}</ansired>"))
            break

        msg = response["message"]
        if response.get("usage"):
            display_usage(response["usage"], elapsed)

        # Display reasoning if present
        if msg.get("reasoning"):
            print_formatted_text(HTML(f"\n<style fg='#888888'>{html.escape(msg['reasoning'][:500])}...</style>"))

        # Add assistant message (preserve reasoning_details for Gemini via OpenRouter)
        assistant_msg = {"role": "assistant", "content": msg.get("content") or ""}
        if msg.get("tool_calls"):
            assistant_msg["tool_calls"] = msg["tool_calls"]
        if msg.get("reasoning"):
            assistant_msg["reasoning"] = msg["reasoning"]
        if msg.get("reasoning_details"):
            assistant_msg["reasoning_details"] = msg["reasoning_details"]
        messages.append(assistant_msg)

        # Display response
        if msg.get("content"):
            print_formatted_text(HTML(f"\n<ansigreen>Assistant:</ansigreen>\n{html.escape(msg['content'])}"))

        # Handle tool calls or finish
        if msg.get("tool_calls"):
            await execute_tools(msg["tool_calls"])
        else:
            break


async def process_turn(user_input: str):
    """Process turn with UI wrapper for cancellation."""
    kb = KeyBindings()
    stop_requested = False
    task = None

    @kb.add("c-w")
    def _(event):
        nonlocal stop_requested
        stop_requested = True

    @kb.add("c-c")
    def _(event):
        if task:
            task.cancel()

    def status():
        if stop_requested:
            return HTML(" <ansiyellow>Stopping...</ansiyellow> (Ctrl+C to force)")
        return HTML(" <ansigreen>Thinking...</ansigreen> (Ctrl+W: stop, Ctrl+C: cancel)")

    app = Application(layout=Layout(Window(FormattedTextControl(status), height=1)), key_bindings=kb)

    async def run():
        nonlocal task
        try:
            await process_turn_logic(user_input, lambda: stop_requested)
        except asyncio.CancelledError:
            print_formatted_text(HTML("\n<ansired>Interrupted.</ansired>"))
        finally:
            app.exit()

    with patch_stdout():
        task = asyncio.create_task(run())
        await app.run_async()
        if not task.done():
            task.cancel()
            try:
                await task
            except asyncio.CancelledError:
                pass


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------


async def main():
    print_formatted_text(HTML(f"<ansicyan>Agent started ({config.mode} mode, {config.model})</ansicyan>"))

    if config.initial_prompt:
        print_formatted_text(HTML("\n<ansicyan>Executing initial prompt...</ansicyan>"))
        await process_turn(config.initial_prompt)

    kb = KeyBindings()

    @kb.add("escape", "z")
    def _(event):
        last_user = next((m["content"] for m in reversed(messages) if m["role"] == "user"), "")
        restored = undo_manager.undo()
        if restored:
            messages[:] = restored
            event.current_buffer.text = last_user if isinstance(last_user, str) else ""
            print_formatted_text(HTML("\n<ansiyellow>Undone last turn.</ansiyellow>"))
        else:
            print_formatted_text(HTML("\n<ansired>Nothing to undo.</ansired>"))

    @kb.add("escape", "enter")
    def _(event):
        event.current_buffer.validate_and_handle()

    @kb.add("escape", "e")
    def _(event):
        event.current_buffer.open_in_editor()

    @kb.add("escape", "i")
    async def _(event):
        path = await save_clipboard_image()
        if path:
            event.current_buffer.insert_text(path)

    session = PromptSession(
        key_bindings=kb,
        vi_mode=True,
        multiline=True,
        lexer=PygmentsLexer(MarkdownLexer),
        style=Style.from_dict({"prompt": "ansicyan bold"}),
    )

    def toolbar():
        mode = session.app.vi_state.input_mode
        labels = {InputMode.INSERT: "INSERT", InputMode.NAVIGATION: "COMMAND", InputMode.REPLACE: "REPLACE"}
        return HTML(f" <b>[{labels.get(mode, mode)}]</b>")

    while True:
        try:
            user_input = await session.prompt_async(
                HTML("<b>User (Alt+Enter: Send, Alt+E: Editor, Alt+I: Image, Alt+Z: Undo, Ctrl+D: Exit):</b>\n"),
                bottom_toolbar=toolbar,
            )
            if user_input.strip().lower() == "exit":
                break
            if user_input.strip():
                await process_turn(user_input)
        except KeyboardInterrupt:
            continue
        except EOFError:
            break


if __name__ == "__main__":
    asyncio.run(main())
